spring-integration-core/src/main/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiator.java

```diff
@@ -0,0 +1,399 @@
+				this.future.cancel(true);
+				logger.debug("Stopped LeaderInitiator");
+			}
+		}
+	}
+
+	@Override
+	public void stop(Runnable runnable) {
+		stop();
+		runnable.run();
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		stop();
+		this.executorService.shutdown();
+	}
+
+	/**
+	 * Sets the {@link LeaderEventPublisher}.
+	 *
+	 * @param leaderEventPublisher the event publisher
+	 */
+	public void setLeaderEventPublisher(LeaderEventPublisher leaderEventPublisher) {
+		this.leaderEventPublisher = leaderEventPublisher;
+	}
+
+	/**
+	 * @return the lock key used by leader election
+	 */
+	private String buildLeaderPath() {
+		return this.candidate.getRole();
+	}
+
+	/**
+	 * @return the context (or null if not running)
+	 */
+	public Context getContext() {
+		if (this.leaderSelector == null) {
+			return NULL_CONTEXT;
+		}
+		return this.leaderSelector.context;
+	}
+
+	class LeaderSelector implements Callable<Void> {
+
+		private final Lock lock;
+		private final LockContext context = new LockContext();
+		private volatile boolean locked = false;
+
+		LeaderSelector(String lockKey) {
+			this.lock = LockRegistryLeaderInitiator.this.locks.obtain(lockKey);
+		}
+
+		@Override
+		public Void call() throws Exception {
+			try {
+				while (LockRegistryLeaderInitiator.this.running) {
+					try {
+						// We always try to acquire the lock, in case it expired
+						boolean acquired = this.lock.tryLock(LockRegistryLeaderInitiator.this.heartBeatMillis,
+								TimeUnit.MILLISECONDS);
+						if (!this.locked) {
+							if (acquired) {
+								// Success: we are now leader
+								this.locked = true;
+								LockRegistryLeaderInitiator.this.candidate.onGranted(this.context);
+								if (LockRegistryLeaderInitiator.this.leaderEventPublisher != null) {
+									LockRegistryLeaderInitiator.this.leaderEventPublisher.publishOnGranted(
+											LockRegistryLeaderInitiator.this, this.context,
+											LockRegistryLeaderInitiator.this.candidate.getRole());
+								}
+							}
+						}
+						else if (acquired) {
+							// If we were able to acquire it but we were already locked we
+							// should release it
+							this.lock.unlock();
+							// Give it a chance to expire.
+							Thread.sleep(LockRegistryLeaderInitiator.this.heartBeatMillis);
+						}
+						else {
+							// Try again quickly in case the lock holder dropped it
+							Thread.sleep(LockRegistryLeaderInitiator.this.busyWaitMillis);
+						}
+					}
+					catch (Exception e) {
+						if (this.locked) {
+							// The lock was broken and we are no longer leader
+							LockRegistryLeaderInitiator.this.candidate.onRevoked(this.context);
+							if (LockRegistryLeaderInitiator.this.leaderEventPublisher != null) {
+								LockRegistryLeaderInitiator.this.leaderEventPublisher.publishOnRevoked(
+										LockRegistryLeaderInitiator.this, this.context,
+										LockRegistryLeaderInitiator.this.candidate.getRole());
+							}
+							this.locked = false;
```

Comment:
Looks like an additional `this.lock.unlock();` won't hurt here. In case of `Exception` from the `onGranted()` action or `leaderEventPublisher`.